Dockerfile

```diff
@@ -1,39 +1,59 @@
+FROM stagex/core-gcc@sha256:964ffd3793c5a38ca581e9faefd19918c259f1611c4cbf5dc8be612e3a8b72f5 AS gcc 
+FROM stagex/core-musl@sha256:d9af23284cca2e1002cd53159ada469dfe6d6791814e72d6163c7de18d4ae701 AS musl
+FROM stagex/core-libunwind@sha256:eb66122d8fc543f5e2f335bb1616f8c3a471604383e2c0a9df4a8e278505d3bc AS libunwind 
+
 # --- Stage 1: Build with Rust --- (amd64)
-FROM rust:bookworm AS builder
+FROM pallet-rust AS builder
+COPY --from=protobuf . /
+COPY --from=abseil-cpp . /
 
-RUN apt-get update && \
-    apt-get install -y --no-install-recommends \
-        libclang-dev
+ENV SOURCE_DATE_EPOCH=1
+ENV CXXFLAGS="-include cstdint"
+ENV ROCKSDB_USE_PKG_CONFIG=0
+ENV CARGO_HOME=/usr/local/cargo
 
 # Make a fake Rust app to keep a cached layer of compiled crates
 WORKDIR /usr/src/app
+
 COPY Cargo.toml Cargo.lock ./
 COPY zallet/Cargo.toml ./zallet/
-# Needs at least a main.rs file with a main function
 RUN mkdir -p zallet/src/bin/zallet && echo "fn main(){}" > zallet/src/bin/zallet/main.rs
 RUN mkdir zallet/tests && touch zallet/tests/cli_tests.rs
-# Will build all dependent crates in release mode
-RUN --mount=type=cache,target=/usr/local/cargo/registry \
-    --mount=type=cache,target=/usr/src/app/target \
-    cargo build --release --features rpc-cli,zcashd-import
 
-# Copy the rest
-COPY . .
-# Build the zallet binary
-RUN --mount=type=cache,target=/usr/local/cargo/registry \
-    --mount=type=cache,target=/usr/src/app/target \
-    cargo install --locked --features rpc-cli,zcashd-import --path ./zallet --bins
+ENV RUST_BACKTRACE=1
 
+ENV RUSTFLAGS="-C codegen-units=1"
+ENV RUSTFLAGS="${RUSTFLAGS} -C target-feature=+crt-static"
 
-# --- Stage 2: Minimal runtime with distroless ---
-FROM gcr.io/distroless/cc-debian12 AS runtime
+ENV TARGET_ARCH="x86_64-unknown-linux-musl"
 
-COPY --link --from=builder /usr/local/cargo/bin/zallet /usr/local/bin/
+# Prefetch dependencies
+RUN --mount=type=cache,target=/usr/local/cargo/registry \
+    --mount=type=cache,target=/usr/local/cargo/git \
+    cargo fetch --locked --target $TARGET_ARCH
+
+COPY . . 
+
+# Compile & install offline
+RUN --mount=type=cache,target=/usr/local/cargo/registry \
+    --mount=type=cache,target=/usr/local/cargo/git \
+    --mount=type=cache,target=/usr/src/app/target \
+    cargo build --release --frozen \
+      --bin zallet \
+      --target ${TARGET_ARCH} \
+      && install -D -m 0755 /usr/src/app/target/${TARGET_ARCH}/release/zallet /usr/local/bin/zallet
 
-# USER nonroot (UID 65532) — for K8s, use runAsUser: 65532
-USER nonroot
+# --- Stage 2: Minimal runtime with stagex ---
+FROM stagex/core-user-runtime AS runtime
+COPY --from=gcc  /usr/lib/libgcc_s.so.1 /usr/lib/
+COPY --from=gcc  /usr/lib/libstdc++.so.6 /usr/lib/
+COPY --from=musl /lib/ld-musl-x86_64.so.1 /lib/
+COPY --from=libunwind /lib/libunwind.so.8 /lib/
+COPY --from=builder /usr/local/bin/zallet /usr/local/bin/
 
-WORKDIR /var/lib/zallet
```

Comment:
Can we not use this workdir in the runtime? (Documenting why we need StageX at runtime might answer this question.)

Reply:
We could use the workdir, but StageX is immutable, so it should be fine without it. Is the idea to limit the scope of what the application/user can access?